src/ripple/protocol/STObject.h

```diff
@@ -327,7 +327,9 @@ class STObject
         v_.reserve (n);
     }
 
-    bool setType (const SOTemplate & type);
+    // Returns an empty string on success.
+    // Returns a non-empty string (describing one error) on error.
+    std::string setType (const SOTemplate & type);
```

Comment:
First, I'm all for changing the name to `applyTemplate()`.  Great suggestion.

Whether returning a `std::string` on error is okay is an excellent question.  And I struggled with the correct way to get the error information out of the call.  I'll go through my thought process and then we can discuss a good solution.

First I considered throwing, as you suggest.  I noted that there are two uses of `setType()` where a throw would break the current expectations.  The one that worried me the most was this `STObject` constructor:  https://github.com/ripple/rippled/blob/develop/src/ripple/protocol/impl/STObject.cpp#L66.  Throwing from `setType()` would turn that into a throwing constructor.  I did not want to research the consequences of that, since my intention was to make a quick fix.  So I removed throwing from my list of options.

I then considered returning a `boost::optional<std::string>`.  I felt that the problem with `optional` is that it usually is non-empty on success.  It's semantics are no better than returning a `std::string`.

What might work well would be returning an Alexandrescu `expected`.  I found that a version of that has gone into boost.  It's called `boost::outcome`.  It is possible that `boost::outcome` would be appropriate.  `boost::outcome` has passed review, but has not yet made it into a boost release.  So it is not yet available to for use.

On a related note, `std::expected` seems to be making its way through the standards committee.  That would fit the desired semantics very well.  But as far as I can tell `std::expected` is targeted at the Library Fundamentals TS, not at the standard.  So I don't think we'll see it soon.  We couldn't possibly see it before C++20 at any rate.

I also considered returning a `std::pair<bool, std::string>`.  But the `bool` would always be in the state opposite to `string.empty()`.  I didn't feel that it provided sufficient additional clarity given the added complexity.

Looking at very big hammers, one could consider returning a `std::exception_ptr` (which is nullable) from `setType()`.  But there is no way to examine an exception's `what()` without causing the `exception_ptr` to throw.  So this would clearly be a bad choice.

Having examined the options, I see the following three choices as viable:

1. Return a `std::string` when as error occurs, which is admittedly a misleading interface.
2. Return a `std::pair<bool, std::string>`, which is less risk prone but more complicated.
3. Throw, in which case I'd better slow down and really study the upstream consequences.

I'm also open to other suggestions.

Thoughts?  Thanks.

Reply:
I believe constructor that you linked is already a throwing constructor via [`STObject::set`](
https://github.com/ripple/rippled/blob/develop/src/ripple/protocol/impl/STObject.cpp#L169) (there are two `set` functions, which do very different things 😡) so I don’t think that’s a big issue.

I like `std::excepted` and think it will allow us to write code that is self-documenting. We can certainly consider than when we get C++20.

My preference would be to `throw`, but you have concerns so I’ll take some time to review the code more carefully because I do respect your judgement.

Reply:
If your preference is to `throw` I'm fine with that.  I'll just need to take some time to understand all of the implications and add any catches upstream that may be required (and there might not be any).